explorer/interpreter/type_checker.cpp

```diff
@@ -1939,43 +1944,54 @@ auto TypeChecker::TypeCheckExp(Nonnull<Expression*> e,
+            op.set_value_category(ValueCategory::Let);
+            return Success();
+          }
+          // Now try an overloaded negation.
+          ErrorOr<Nonnull<Expression*>> result = BuildBuiltinMethodCall(
+              impl_scope, op.arguments()[0],
+              BuiltinInterfaceName{Builtins::Negate}, BuiltinMethodCall{"Op"});
+          if (!result.ok()) {
+            // We couldn't find a matching `impl`.
+            return CompilationError(e->source_loc())
+                   << "type error in `" << ToString(op.op()) << "`:\n"
+                   << result.error().message();
+          }
+          op.set_rewritten_form(*result);
           return Success();
+        }
         case Operator::Add:
-          CARBON_RETURN_IF_ERROR(ExpectExactType(e->source_loc(), "addition(1)",
-                                                 arena_->New<IntType>(), ts[0],
-                                                 impl_scope));
-          CARBON_RETURN_IF_ERROR(ExpectExactType(e->source_loc(), "addition(2)",
-                                                 arena_->New<IntType>(), ts[1],
-                                                 impl_scope));
-          op.set_static_type(arena_->New<IntType>());
-          op.set_value_category(ValueCategory::Let);
-          return Success();
         case Operator::Sub:
-          CARBON_RETURN_IF_ERROR(
-              ExpectExactType(e->source_loc(), "subtraction(1)",
-                              arena_->New<IntType>(), ts[0], impl_scope));
-          CARBON_RETURN_IF_ERROR(
-              ExpectExactType(e->source_loc(), "subtraction(2)",
-                              arena_->New<IntType>(), ts[1], impl_scope));
-          op.set_static_type(arena_->New<IntType>());
-          op.set_value_category(ValueCategory::Let);
-          return Success();
-        case Operator::Mul:
-          CARBON_RETURN_IF_ERROR(
-              ExpectExactType(e->source_loc(), "multiplication(1)",
-                              arena_->New<IntType>(), ts[0], impl_scope));
-          CARBON_RETURN_IF_ERROR(
-              ExpectExactType(e->source_loc(), "multiplication(2)",
-                              arena_->New<IntType>(), ts[1], impl_scope));
-          op.set_static_type(arena_->New<IntType>());
-          op.set_value_category(ValueCategory::Let);
+        case Operator::Mul: {
+          // Handle a built-in operator first.
+          if (isa<IntType>(ts[0]) && isa<IntType>(ts[1]) &&
+              IsSameType(ts[0], ts[1], impl_scope)) {
+            op.set_static_type(ts[0]);
+            op.set_value_category(ValueCategory::Let);
+            return Success();
+          }
+          // Now try an overloaded operator.
+          Builtins::Builtin builtin =
+              op.op() == Operator::Add   ? Builtins::AddWith
+              : op.op() == Operator::Sub ? Builtins::SubWith
+                                         : Builtins::MulWith;
```

Comment:
I kind of feel like this should be a switch; it feels like it gets worse to read as you add more (like DivWith, ModWith, etc).

Reply:
Factored out the case body and integrated this into the outer `switch`.